### PR TITLE
[tp160065] - Final steps toward v3 support

### PR DESCRIPTION
**Pull request by ndurell**
# Description
This adds support for the latest Klaviyo APIs (2022-10-17). 
This should be one of the last PRs before the next release.

Included in here:

* centralized store for managing state changes
* use of v3 for creating profiles
* use of v3 for creating events
* use of v1 POST for saving tokens
* semi-persistent queue of individual requests
* batched triggering of requests (adjusted basic on network changes).
* vendorized some of Composable Architecture lib for state management
* some included (test-only) dependencies to support the TestStore
* updates to the Klaviyo class to call into the new store.

# TODOS
Mostly testing

1. more coverage on the state management file.
2. test coverage on the klaviyo class
3. manually test open push
4. maybe a better test that involves cancellation.

### CODE REVIEW

File: Sources/KlaviyoSwift/InternalAPIModels.swift [hunk 1]
```diff
@@ -30,11 +31,11 @@ extension KlaviyoAPI.KlaviyoRequest {
                     let properties: AnyCodable
                     enum CodingKeys: String, CodingKey {
                         case email
-                        case phoneNumber
-                        case externalId
-                        case anonymousId
-                        case firstName
-                        case lastName
+                        case phoneNumber = "phone_number"
```
Comment (ndurell) [inline Sources/KlaviyoSwift/InternalAPIModels.swift:34]: I ran into some encoding issues with this request and codeable and had to fix it up.

File: Sources/KlaviyoSwift/KlaviyoEnvironment.swift [hunk 2]
```diff
@@ -13,14 +13,12 @@ var environment = KlaviyoEnvironment.production
 let PRODUCTION_HOST = "https://a.klaviyo.com"
 let encoder = {
     let encoder = JSONEncoder()
-    encoder.keyEncodingStrategy = .convertToSnakeCase
```
Comment (ndurell) [inline Sources/KlaviyoSwift/KlaviyoEnvironment.swift:16]: 🐍 case was biting me so I turned it off. Everything is manually specified now.

File: Sources/KlaviyoSwift/Vendor/ComposableArchitecture/Cancellation.swift [hunk 3]
```diff
@@ -0,0 +1,309 @@
+//
```
Comment (ndurell) [inline Sources/KlaviyoSwift/Vendor/ComposableArchitecture/Cancellation.swift:1]: This is vendor code from https://github.com/pointfreeco/swift-composable-architecture. I mostly left it as is except for excluding dependencies. I'm going to add a todo with a readme on why this is here and not fully included. Potentially in the future we can have this as a separate target or even package.

File: Tests/KlaviyoSwiftTests/__Snapshots__/EncodableTests/testKlaviyoState.1.json [hunk 4]
```diff
@@ -13,32 +11,17 @@
           "_0" : {
             "properties" : {
               "$anonymous" : "foo",
+              "$append" : {
```
Comment (ndurell) [inline Tests/KlaviyoSwiftTests/__Snapshots__/EncodableTests/testKlaviyoState.1.json:14]: messed this up before. This uses a weird append feature.

File: Tests/KlaviyoSwiftTests/StateManagementTests.swift [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+                                        anonymousId: environment.analytics.uuid().uuidString,
+                                        pushToken: "blob_token",
+                                        queue: [],
+                                        requestsInFlight: [],
+                                        initialized: true,
+                                        flushing: false)
+        let request = try initialState.buildProfileRequest()
+        let request2 = try initialState.buildTokenRequest()
+        initialState.queue = [request, request2]
+        let store = TestStore(initialState: initialState, reducer: KlaviyoReducer())
```
Comment (ndurell) [inline Tests/KlaviyoSwiftTests/StateManagementTests.swift:231]: These tests walk you through state changes in the store. I have them set for exhaustive mode which means you have to spell out each change that occurs as well every action that is fired.

File: Tests/KlaviyoSwiftTests/__Snapshots__/EncodableTests/testEventPayload.1.json [hunk 6]
```diff
@@ -3,7 +3,7 @@
     "attributes" : {
       "metric" : {
         "name" : "blob",
-        "service" : "ios-analytics"
+        "service" : "klaviyo"
```
Comment (ndurell) [inline Tests/KlaviyoSwiftTests/__Snapshots__/EncodableTests/testEventPayload.1.json:6]: I changed this back to klaviyo so it shows up properly in the UI. We might think about getting our icon to indicate this is coming from ios but for now this should be left as is.

File: Sources/KlaviyoSwift/StateManagement.swift [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+            }
+            return .run { send in
+                let result = await environment.analytics.klaviyoAPI.send(request)
+                switch result {
+                case .success(_):
+                    // TODO: may want to inspect response further.
+                    await send(.dequeCompletedResults(request))
+                case .failure(_):
+                    // TODO: depending on failure may want to deque
+                    await send(KlaviyoAction.cancelInFlightRequests)
```
Comment (ndurell) [inline Sources/KlaviyoSwift/StateManagement.swift:171]: I probably should limit retries here...

File: Sources/KlaviyoSwift/StateManagement.swift [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+                saveKlaviyoState(state: state)
+            }
+ 
+        case .enqueueLegacyEvent(let legacyEvent):
+            // TODO: Needs a few tests.
+            guard let apiKey = state.apiKey else {
+                return .none
+            }
+            // TODO: might need to update state based on data in here.
+            return .run { [state] send in
```
Comment (ndurell) [inline Sources/KlaviyoSwift/StateManagement.swift:217]: Add a limit on how many requests can be queued at a time.

File: Sources/KlaviyoSwift/StateManagement.swift [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+                environment.logger.error("request error")
+                // TODO: maybe better handling here...
+            }.cancellable(id: RequestId.self)
+        case .cancelInFlightRequests:
+            state.flushing = false
+            state.queue.insert(contentsOf: state.requestsInFlight, at: 0)
+            state.requestsInFlight = []
+            return .none
+        case .networkConnectivityChanged(let networkStatus):
+            guard state.initialized else {
```
Comment (ndurell) [inline Sources/KlaviyoSwift/StateManagement.swift:183]: This allows us to adjust flush interval based on the network. For now I'm using 10 seconds on wifi and 30 seconds on cellular.

File: Sources/KlaviyoSwift/StateManagement.swift [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+                }))
+        case .start:
+            guard state.initialized else {
+                return .none
+            }
+            return environment.analytics.timer(state.flushInterval)
+                    .map { _ in
+                        KlaviyoAction.flushQueue
+                    }.eraseToEffect()
+                    .cancellable(id: Timer.self, cancelInFlight: true)
```
Comment (ndurell) [inline Sources/KlaviyoSwift/StateManagement.swift:142]: These cancellables allow us to cancel inflight requests for example when the app goes to the background.

File: Sources/KlaviyoSwift/KlaviyoEnvironment.swift (hunk 2 above)
Comment (evan-masseau) [reply Sources/KlaviyoSwift/KlaviyoEnvironment.swift:16]: mess with the snake you're gonna get the fangs

File: Sources/KlaviyoSwift/StateManagement.swift (hunk 9 above)
Comment (evan-masseau) [reply Sources/KlaviyoSwift/StateManagement.swift:183]: nice

File: Sources/KlaviyoSwift/StateManagement.swift (hunk 7 above)
Comment (jordan-griffin) [reply Sources/KlaviyoSwift/StateManagement.swift:171]: Right now, if someone's network drops and the request fails is it just gonna attempt resending indefinitely?

File: Sources/KlaviyoSwift/Klaviyo.swift [hunk 11]
```diff
@@ -220,63 +190,16 @@ public class Klaviyo: NSObject  {
      */
     public func trackEvent(event: String?, customerProperties: NSDictionary?, propertiesDict: NSDictionary?, eventDate: NSDate?) {
         
-        var eventName = event
-        // Set default track name if none provided
-        if (eventName == nil || eventName!.isEmpty) { eventName = "KL_Event" }
-
+        guard let eventName = event, !eventName.isEmpty else {
```
Comment (evan-masseau) [inline Sources/KlaviyoSwift/Klaviyo.swift:193]: why is event an optional in the first place? Is that just backward compatibility?

File: Sources/KlaviyoSwift/Klaviyo.swift [hunk 12]
```diff
@@ -464,288 +336,28 @@ public class Klaviyo: NSObject  {
...
-            let objects = obj as! NSDictionary
-            
-            for key in objects.keyEnumerator() {
-                guard let stringKey = key as? String else {
-                    print("warning: property keys should be strings. got: \(String(describing: key))")
-                    continue
-                }
-                let v : AnyObject = JSONSerializableObjectForObject(obj: objects[stringKey]! as AnyObject)
-                dict[stringKey] = v
+    private func dispatchOnMainThread(action: KlaviyoAction) {
```
Comment (evan-masseau) [inline Sources/KlaviyoSwift/Klaviyo.swift:355]: Sorry if I missed an explanation somewhere, but just curious what led to needing to specify main thread for some actions?

**Review by ajaysubra — COMMENTED** [review]
Briefly looked at this PR and had some questions and posting them here. Cool use of async await and TCA.

File: Sources/KlaviyoSwift/Klaviyo.swift [hunk 13]
```diff
@@ -113,29 +104,20 @@ public class Klaviyo: NSObject  {
...
      - Parameter apiKey: string representation of the Klaviyo API Key
      */
-    public class func setupWithPublicAPIKey(apiKey : String) {
-        sharedInstance.apiKey = apiKey
-        sharedInstance.unarchive()
-        sharedInstance.flush()
-        sharedInstance.addNotificationObserver()
+    public class func setupWithPublicAPIKey(apiKey: String) {
+        //_ avoids warning from xcode
+        sharedInstance.dispatchOnMainThread(action: .initialize(apiKey))
```
[the remark this answers is not in the document]
Comment (ndurell) [reply Sources/KlaviyoSwift/Klaviyo.swift:120]: The store is designed to only work on the main thread. I think eventually we could move to our thread. Also to be clear this doesn't mean that store operations are done on the main thread. The side effects still occur on the background. Just if we are updating the state it is done on the main thread.

File: Sources/KlaviyoSwift/StateManagement.swift [hunk 14]
```diff
@@ -0,0 +1,293 @@
...
+    typealias Action = KlaviyoAction
+    func reduce(into state: inout KlaviyoSwift.KlaviyoState, action: KlaviyoSwift.KlaviyoAction) -> EffectTask<KlaviyoSwift.KlaviyoAction> {
+        switch action {
+        case let .initialize(apiKey):
+            if state.initialized {
+                return .none
+            }
+            state.apiKey = apiKey
+            return .run { send in
+                let initialState = loadKlaviyoStateFromDisk(apiKey: apiKey)
```
[the remark this answers is not in the document]
Comment (ndurell) [reply Sources/KlaviyoSwift/StateManagement.swift:55]: potentialy but my code should handle that case.

File: Sources/KlaviyoSwift/StateManagement.swift [hunk 15]
```diff
@@ -0,0 +1,293 @@
...
+                profile: .init(attributes: .init(
+                    email: email,
+                    phoneNumber: phoneNumber,
+                                            externalId: externalId,
+                                            properties: properties)
+                ),
+                anonymousId: anonymousId)
+        )
+        let endpoint = KlaviyoAPI.KlaviyoRequest.KlaviyoEndpoint.createProfile(payload)
+        guard let apiKey = apiKey else {
```
[the remark this answers is not in the document]
Comment (ndurell) [reply Sources/KlaviyoSwift/StateManagement.swift:260]: Well technically I made it non-optional in the Klaviyo api class. Technically checking is initialized should be enough. I could probably clean up this section a bit.

**Comment by evan-masseau** [comment]
My overall feedback is this looks like good steps forward, but its a codebase I'm not super familiar with anyway. I'm happy to see this merged and continue to learn about what you've done as I get more time to work on iOS.

File: Sources/KlaviyoSwift/Klaviyo.swift (hunk 12 above)
Comment (ndurell) [reply Sources/KlaviyoSwift/Klaviyo.swift:355]: The store implementation I'm using requires main thread. Eventually I think we can move it to our own thread but the intention is that none of the non-side effect is very heavy weight so this should be ok.

Comment (ndurell) [reply Sources/KlaviyoSwift/Klaviyo.swift:355]: This is just a protection against someone calling our sdk from another thread.

**Review by jordan-griffin — APPROVED** [review]

**Pull request merged**